src/adapters/source-adapters/CoinbaseOracleSourceAdapter.sol

```diff
@@ -0,0 +1,99 @@
+     * @param timestamp The timestamp to try getting the latest data at.
+     * @param maxTraversal The maximum number of rounds to traverse when looking for historical data.
+     * @return answer The answer as of the requested timestamp, or the earliest available data if not available, in 18 decimals.
+     * @return updatedAt The timestamp of the answer.
+     */
+    function tryLatestDataAt(uint256 timestamp, uint256 maxTraversal)
+        public
+        view
+        virtual
+        override
+        returns (int256, uint256)
+    {
+        (int256 answer, uint256 updatedAt) = _tryLatestRoundDataAt(timestamp, maxTraversal);
+        return (DecimalLib.convertDecimals(answer, SOURCE_DECIMALS, 18), updatedAt);
+    }
+
+    /**
+     * @notice Initiate a snapshot of the source data. This is a no-op for Coinbase.
+     */
+    function snapshotData() public virtual override {}
+
+    /**
+     * @notice Returns the latest data from the source.
+     * @return answer The latest answer in 18 decimals.
+     * @return updatedAt The timestamp of the answer.
+     */
+    function getLatestSourceData() public view virtual override returns (int256, uint256) {
+        (, int256 sourceAnswer,, uint256 updatedAt,) = COINBASE_SOURCE.latestRoundData();
+        return (DecimalLib.convertDecimals(sourceAnswer, SOURCE_DECIMALS, 18), updatedAt);
+    }
+
+    // Tries getting the latest data as of the requested timestamp. If this is not possible,
+    // returns the earliest data available past the requested timestamp considering the maxTraversal limitations.
+    function _tryLatestRoundDataAt(uint256 timestamp, uint256 maxTraversal) internal view returns (int256, uint256) {
+        (uint80 roundId, int256 answer,, uint256 updatedAt,) = COINBASE_SOURCE.latestRoundData();
+
+        // If the latest update is older than or equal to the requested timestamp, return the latest data.
+        if (updatedAt <= timestamp) {
+            return (answer, updatedAt);
+        }
+
+        // Attempt traversing historical round data backwards from roundId.
+        (int256 historicalAnswer, uint256 historicalUpdatedAt) = _searchRoundDataAt(timestamp, roundId, maxTraversal);
+
+        // Validate returned data. If it is uninitialized, fall back to returning the current latest round data.
+        if (historicalUpdatedAt > 0) {
+            return (historicalAnswer, historicalUpdatedAt);
+        }
+
+        return (answer, updatedAt);
+    }
+
+    // Searches for the latest historical data not newer than the requested timestamp.
+    // Returns newer data than requested if it exceeds traversal limits or holds uninitialized data that should be handled by the caller.
+    function _searchRoundDataAt(uint256 timestamp, uint80 latestRoundId, uint256 maxTraversal)
+        internal
+        view
+        returns (int256, uint256)
+    {
+        int256 answer;
+        uint256 updatedAt;
```

Comment:
These could be named return props then they don't need to be defined again. in some places we do that with named return props. might save a small bit of gas as well.